cpp/src/arrow/dataset/dataset.h

```diff
@@ -134,6 +135,8 @@ class ARROW_DS_EXPORT InMemoryFragment : public Fragment {
 
 /// @}
 
+using FragmentGenerator = AsyncGenerator<std::shared_ptr<Fragment>>;
```

Comment:
The following avoids having to include `async_generator.h`
```suggestion
using FragmentGenerator = std::function<Future<std::shared_ptr<Fragment>>()>;
```

Reply:
Actually I would want to introduce something like `async_generator_fwd.h` and place the main definitions (both `using`:s and class definitions like `PushGenerator` etc.) there. Just in case `AsyncGenerator` becomes something more elaborate than a plain `std::function`, your approach would require a lot of patching "here and there".